### PR TITLE
[rhoai-2.25] fix(build): derive native ONNX/pyarrow versions from pylock.toml

### PR DESCRIPTION
**Pull request by jiridanek**
## Description

Replace hardcoded `ARG ONNX_VERSION` / `ARG PYARROW_VERSION` with a stdlib-only `scripts/pylock_version.py` helper that reads pinned versions from each workbench `pylock.toml` using PEP 508-style platform markers.

**Scope (images touched):**
- `jupyter/datascience`, `runtimes/datascience` — ppc64le ONNX + pyarrow native builds; s390x pyarrow
- `jupyter/trustyai` — `devel_env_setup.sh` ppc64le wheel builds (torch/pyarrow/pillow/accelerate)

**`scripts/pylock_version.py`:**
- Stdlib-only (`tomllib` + `ast` marker evaluator); runs in UBI `python-312` before `uv` is installed
- Prints raw PEP 440 versions; shell applies prefixes (`v${ONNX_VERSION}`, `apache-arrow-${PYARROW_VERSION}`, `${TORCH_VERSION%%+*}` for git tags)
- Marker evaluation via `ast.parse` — supports nested `and`/`or`/`not`, `in` tuples, `.*` wildcards; rejects unsupported operators loudly

**Also in this PR:**
- ONNX **1.22.0** (from pylock) removed upstream `requirements.txt` — ppc64le `onnx-builder` uses `requirements-min.txt`
- Trustyai `devel_env_setup.sh` shell quoting + unset-safe `LD_LIBRARY_PATH` under `set -u`
- Semgrep `shell-unquoted-var-in-dangerous-cmd` hardened (word boundaries; stop at `|`, `||`, `;`, `&`, `&&`)
- `tests/with_subtests.py` decorator for iterable-driven subtests (quieter alternative to manual `for`/`with subtests.test`) — tracks [opendatahub-io/notebooks#4140](https://github.com/opendatahub-io/notebooks/issues/4140)

**Commits (5):**
1. `feat(scripts): add stdlib pylock_version helper`
2. `fix(build): derive native ONNX/pyarrow versions from pylock.toml`
3. `fix(trustyai): resolve pins via pylock and quote shell vars`
4. `fix(semgrep): harden shell-unquoted-var-in-dangerous-cmd`
5. `test: add with_subtests helper for iterable static checks`

## How Has This Been Tested?

- [x] `uv run pytest` on `tests/unit/scripts/test_pylock_version.py`, `tests/test_native_build_versions.py`, `tests/unit/test_subtests.py`, `tests/unit/test_semgrep_shell_unquoted.py`
- [x] `python3.12 -S scripts/pylock_version.py` smoke test on host and in `registry.access.redhat.com/ubi9/python-312:latest`
- [x] `uv run prek --all-files` (ruff green; pre-existing pyright noise on unrelated `runtime_test.py`)
- [ ] Konflux: jupyter/runtime datascience ppc64le after merge (openblas follow-up is #2534)

Self checklist:
- [x] Targeted unit tests run before review
- [x] Konflux-specific changes are in `Dockerfile.konflux.*` under `rhds/notebooks` as required for rhoai-2.25

## Merge criteria

- [x] Focused thematic commits (not one mega-squash)
- [x] Testing instructions in PR body
- [ ] Review / lgtm / approve

## Out of scope / follow-ups

- [#2534](https://github.com/red-hat-data-services/notebooks/pull/2534) — `openblas-devel` in `onnx-builder` for ppc64le numpy sdist (stacked after this merges)
- Broader adoption of `with_subtests` across the suite — [ODH #4140](https://github.com/opendatahub-io/notebooks/issues/4140)
- whl-cache dedup (trustyai double `uv pip install`)

## Summary by CodeRabbit

* **Improvements**
  * Native CPU and runtime images now derive ONNX and PyArrow versions from the lockfile, producing matching architecture-specific tags (including ppc64le and s390x).
  * TrustyAI wheel caching now includes the version-resolution helper for consistent builds.
  * ONNX native builds now install from a minimal dependency set.
* **Build/Configuration**
  * Added lock-aware `pylock_version` tooling and use it during Docker builds and development setup (with a configurable `DATASCIENCE_SOURCE_CODE` build argument).
* **Tests**
  * Added coverage validating Dockerfiles and marker-based version resolution follow the lockfile, plus new unit tests for the helper, Semgrep rule, and subtest behavior.
* **Chores**
  * Reduced noisy pytest subtest output and tightened the Semgrep shell-variable detection rule.

### CODE REVIEW

File: runtimes/datascience/ubi9-python-3.12/Dockerfile.cpu [hunk 1]
```diff
@@ -58,13 +61,16 @@ EOF
...
 export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig/
-export LD_LIBRARY_PATH=/usr/local/lib64:/usr/local/lib:/usr/lib64:/usr/lib${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}
+export LD_LIBRARY_PATH=/usr/local/lib64:/usr/local/lib:/usr/lib64:/usr/lib\${LD_LIBRARY_PATH:+:\$LD_LIBRARY_PATH}
 export OPENBLAS_VERSION=0.3.30
-export ONNX_VERSION=1.20.1
-export PYARROW_VERSION=17.0.0
-export PATH="$HOME/.cargo/bin:$PATH"
+export ONNX_VERSION=${ONNX_VERSION}
+export PYARROW_VERSION=${PYARROW_VERSION}
+export PATH="\$HOME/.cargo/bin:\$PATH"
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply runtimes/datascience/ubi9-python-3.12/Dockerfile.cpu:74]: @coderabbitai recheck, rewritten

File: scripts/pylock_version.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        "platform_python_implementation",
+        "sys_platform",
+        "platform_machine",
+        "platform_system",
+        "platform_release",
+        "platform_version",
+        "os_name",
+        "extra",
+    }
+)
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply scripts/pylock_version.py:45]: @coderabbitai recheck

**Comment by jiridanek** [comment]
@coderabbitai Code Review: PR #2519 — [rhoai-2.25] fix(build): derive native ONNX/pyarrow versions from pylock.toml
     
  Overview
     
  Replaces hardcoded ARG ONNX_VERSION=v1.20.1 / ARG PYARROW_VERSION=17.0.0 across 8 Dockerfiles (4 ODH .cpu + 4 Konflux .konflux.cpu) with a new stdlib-only scripts/pylock_version.py that reads pinned versions from each workbench's pylock.toml. Also adds a PEP 508 marker evaluator, ONNX 1.22.0
  requirements-min.txt fix, and shell quoting improvements in devel_env_setup.sh.

  11 files changed, +465 / -51.

  ---
  Positives
  
  - Single source of truth: Eliminates the version-drift problem where Dockerfile ARGs and pylock pins diverge after lock regeneration.
  - Stdlib-only script: pylock_version.py uses only tomllib + ast, so it runs in bare UBI python-312 before uv is installed — correct for the build stage ordering.
  - Solid marker evaluator: The ast.parse approach with an explicit _ALLOWED_NAMES whitelist and ValueError for unknown variables/operators is defensive and auditable. Much better than regex-grepping pylock.toml.
  - Good test coverage: 19 tests covering CLI invocation modes, marker edge cases, missing packages, and Dockerfile-aware integration checks.
  - Shell quoting fixes: "${ONNX_TAG}", "${TORCH_TAG}", quoted WHEEL_DIR in trustyai/codeserver address real semgrep findings.

  ---
  Issues
  
  1. --platform x86_64 inside a ppc64le block (confusing, not wrong)

  devel_env_setup.sh:138 and :171:
  if [[ $(uname -m) == "ppc64le" ]]; then
      TORCH_VERSION=$(python3 ./pylock_version.py torch --platform x86_64)
      ...
      uv pip install ... "accelerate==$(python3 ./pylock_version.py accelerate --platform x86_64)"
  These resolve the x86_64 version because torch/accelerate only have x86_64 markers in the pylock (no ppc64le wheel exists — it's built from source). This is functionally correct but a maintenance trap: a future reader will think it's a copy-paste bug. A one-line comment explaining "torch is pinned under x86_64
  markers; we build the same version from source on ppc64le" would prevent a well-intentioned "fix."

  2. Unquoted ${PYARROW_VERSION} in git clone (pre-existing, but newly constructed)

  devel_env_setup.sh:155:
  git clone --recursive https://github.com/apache/arrow.git -b apache-arrow-${PYARROW_VERSION}
  ${PILLOW_VERSION} at line 164 and ${WHEELS_DIR}/*.whl at line 171 are also unquoted. The torch one was correctly quoted as "${TORCH_TAG}". Since this PR already fixed quoting for TORCH_TAG, it would be consistent to quote the remaining variable expansions in the same block. Word splitting on a version string is
  unlikely but free to prevent. 

  3. Heredoc escaping in runtime-datascience profile script is correct but fragile

  runtimes/datascience/.../Dockerfile.cpu:197-208: The heredoc switched from <<'PROFILE_EOF' (literal) to <<PROFILE_EOF (expanding) to inject ${ONNX_VERSION} and ${PYARROW_VERSION}. The \$ escapes on LD_LIBRARY_PATH and PATH are correct. However, adding any new $VAR line to this profile script in the future will
  silently expand at build time instead of at login time. Consider a comment at the heredoc marker: # NB: unquoted heredoc — escape $ for vars that must expand at login.

  4. re.DOTALL false positive in test assertion

  tests/test_native_build_versions.py:33:
  lookup = re.compile(rf"pylock_version\.py.*\b{package}\b", re.DOTALL)
  re.DOTALL makes . match newlines, so this succeeds if pylock_version.py appears on line 10 and onnx appears on line 200 in any unrelated context (e.g., a pip install onnx comment). CodeRabbit flagged this too. Dropping re.DOTALL (or using their single-line regex) would make the assertion meaningful.
  
  5. conftest.py fixture uses importlib — consider sys.path insert instead

  @pytest.fixture(scope="session")
  def pylock_version():
      spec = importlib.util.spec_from_file_location("pylock_version", _SCRIPTS / "pylock_version.py")
      ...
  This works but means pylock_version isn't importable the normal way anywhere else in the test suite. Since scripts/ isn't a package (no __init__.py), this is the right call. But if more scripts get tested this way, a conftest that does sys.path.insert(0, str(_SCRIPTS)) once would be simpler. Not blocking — just
  a style note for when it recurs.

  6. test_pylock_pins_resolve_for_native_arches asserts pyarrow_ppc64le == pyarrow_s390x

  tests/test_native_build_versions.py:40:
  assert pyarrow_ppc64le == pyarrow_s390x
  This enforces that both arches pin the same pyarrow version. If they ever diverge intentionally (e.g., s390x needs a patched version), this test breaks. Consider whether this is an intended contract or an incidental coincidence — if incidental, remove it.

  7. python_minor_from_path fallback to "3.12" is silent

  scripts/pylock_version.py:49:
  return match.group(1) if match else "3.12"
  If the pylock path doesn't contain python-X.Y, this silently defaults to 3.12. For a lockfile at /tmp/pylock.toml (as used in Dockerfiles), the path won't match — so the default is always used. This is fine today (everything is 3.12) but could mask a real mismatch when 3.13 arrives. A --python CLI flag exists
  as an override, which mitigates this. Low priority. 

  8. Minor: _parse_args metavar is misleading

  metavar=("PACKAGE", "PYLOCK"),
  The actual order is PYLOCK PACKAGE (pylock first, package second) per the code at line 178-179. The metavar tuple should be ("PYLOCK", "PACKAGE") to match the positional semantics. Currently --help shows them backwards.

  ---
  Security

  - No injection risk: TARGETARCH comes from Docker's constrained ARG; pylock paths are build-context-local.
  - The ast.parse(marker, mode="eval") approach is safe — eval mode only parses expressions, never exec/import. The _eval_marker function never calls eval() on the AST.

  ---
  Summary
  
  Well-structured change that solves a real version-drift problem. The marker evaluator is solid, the test coverage is good, and the Dockerfile changes are mechanically correct. The main actionable items are:

  1. Add a comment explaining the --platform x86_64 usage inside ppc64le blocks in devel_env_setup.sh
  2. Drop re.DOTALL from the test regex to avoid false positives
  3. Fix the metavar order in _parse_args
  4. Consider quoting the remaining ${PYARROW_VERSION} and ${PILLOW_VERSION} expansions for consistency

File: jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh [hunk 3]
```diff
@@ -48,9 +48,12 @@ if [[ $(uname -m) == "ppc64le" ]]; then
 
     # Torch
     cd ${CURDIR}
-    TORCH_VERSION=$(grep -A1 '"torch"' pylock.toml | grep -Eo '\b[0-9\.]+\b')
+    TORCH_VERSION=$(python3 ./pylock_version.py torch --platform x86_64)
```
Comment (jiridanek) [inline jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh:53]: This resolves the x86_64 marker version inside a `ppc64le` block. It's correct — torch only has x86_64 markers in the pylock (no ppc64le wheel exists, it's built from source) — but a future reader will think it's a copy-paste bug.

A one-line comment would prevent a well-intentioned "fix":
```bash
# torch is pinned under x86_64 markers; we build the same version from source on ppc64le
TORCH_VERSION=$(python3 ./pylock_version.py torch --platform x86_64)
```

Same applies to the `accelerate` call on line 101.

File: jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh [hunk 4]
```diff
@@ -61,7 +64,7 @@ if [[ $(uname -m) == "ppc64le" ]]; then
 
     cd ${CURDIR}
     # Pyarrow
-    PYARROW_VERSION=$(grep -A1 '"pyarrow"' pylock.toml | grep -Eo '\b[0-9\.]+\b')
+    PYARROW_VERSION=$(python3 ./pylock_version.py pyarrow --platform ppc64le)
```
Comment (jiridanek) [inline jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh:69]: This PR already fixed quoting for `TORCH_TAG` (nice), but `PYARROW_VERSION` here is still unquoted. Same for `PILLOW_VERSION` on line 90. Word splitting on a version string is unlikely but free to prevent:
```bash
git clone --recursive https://github.com/apache/arrow.git -b "apache-arrow-${PYARROW_VERSION}"
```

File: runtimes/datascience/ubi9-python-3.12/Dockerfile.cpu [hunk 5]
```diff
@@ -58,13 +61,16 @@ EOF
 
 RUN /bin/bash <<'EOF'
 set -Eeuxo pipefail
-if [ "$TARGETARCH" = "ppc64le" ]; then cat > /etc/profile.d/ppc64le.sh <<'PROFILE_EOF'
+if [ "$TARGETARCH" = "ppc64le" ]; then
+    ONNX_VERSION=$(python3 /tmp/pylock_version.py /tmp/pylock.toml onnx --platform ppc64le)
+    PYARROW_VERSION=$(python3 /tmp/pylock_version.py /tmp/pylock.toml pyarrow --platform ppc64le)
+    cat > /etc/profile.d/ppc64le.sh <<PROFILE_EOF
```
Comment (jiridanek) [inline runtimes/datascience/ubi9-python-3.12/Dockerfile.cpu:68]: The heredoc switched from `<<'PROFILE_EOF'` (literal, no expansion) to `<<PROFILE_EOF` (expanding) so `${ONNX_VERSION}` and `${PYARROW_VERSION}` are injected at build time. The `\$` escapes on `LD_LIBRARY_PATH` and `PATH` are correct.

However, this is a subtle trap: adding any new `$VAR` line to this profile script in the future will silently expand at build time instead of at login time. Consider a comment at the heredoc marker:
```dockerfile
    # NB: unquoted heredoc — escape $ for vars that must expand at login
    cat > /etc/profile.d/ppc64le.sh <<PROFILE_EOF
```

File: tests/test_native_build_versions.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        re.compile(r"apache-arrow-17\.0\.0"),
+    )
+    for dockerfile in dockerfiles:
+        text = dockerfile.read_text()
+        for pattern in forbidden:
+            assert not pattern.search(text), f"{dockerfile.name} still uses {pattern.pattern}"
+        assert "pylock_version.py" in text, f"{dockerfile.name} should copy pylock_version.py"
+        for package in ("onnx", "pyarrow"):
+            lookup = re.compile(rf"pylock_version\.py.*\b{package}\b", re.DOTALL)
+            assert lookup.search(text), f"{dockerfile.name} should resolve {package} via pylock_version.py"
```
Comment (jiridanek) [inline tests/test_native_build_versions.py:47]: `re.DOTALL` makes `.` match newlines, so this succeeds if `pylock_version.py` appears on line 10 and `onnx` appears on line 200 in any unrelated context (e.g., a `pip install onnx` comment). Drop `re.DOTALL` to make the assertion meaningful:
```python
lookup = re.compile(rf"pylock_version\.py\s+(?:\S+\s+)?{package}\b")
```

File: tests/test_native_build_versions.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+
+@pytest.mark.parametrize("image_dir", NATIVE_IMAGE_DIRS, ids=lambda p: p.relative_to(ROOT).as_posix())
+def test_pylock_pins_resolve_for_native_arches(image_dir: Path, pylock_version) -> None:
+    pylock = image_dir / "pylock.toml"
+    locked_version = pylock_version.locked_version
+    onnx_ppc64le = locked_version(pylock, "onnx", platform_machine="ppc64le")
```
Comment (jiridanek) [inline tests/test_native_build_versions.py]: This enforces that ppc64le and s390x pin the same pyarrow version. Is this an intended contract or an incidental coincidence? If they ever diverge intentionally (e.g., s390x needs a patched version), this test breaks. Consider removing this assertion or adding a comment explaining the invariant.

File: scripts/pylock_version.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        pylock_path.read_text(),
+        python_minor=resolved_python,
+        platform_machine=platform_machine,
+    )
+    if package not in packages:
+        raise LookupError(
+            f"package {package!r} not found in {pylock_path} "
+            f"for platform_machine={platform_machine!r} python={resolved_python!r}"
+        )
+    return packages[package]["version"]
```
Comment (jiridanek) [inline scripts/pylock_version.py:169]: Minor: the metavar tuple says `("PACKAGE", "PYLOCK")` but the actual positional order is `PYLOCK PACKAGE` (pylock first when two args are given, per the `match` at line 178). This makes `--help` show them backwards:
```
positional arguments:
  PACKAGE     ...
  PYLOCK      ...
```
Should be `metavar=("PYLOCK", "PACKAGE")` or `metavar=("ARG",)`.

**Review by jiridanek — COMMENTED** [review]
Nice change — single source of truth for native build versions is a real improvement. The marker evaluator is solid and the `ast.parse` approach is safe.

Main actionable items in the inline comments:
1. **Comment the `--platform x86_64` inside ppc64le blocks** — correct but looks like a bug to future readers
2. **Drop `re.DOTALL`** from the test regex — currently matches across unrelated lines
3. **Fix metavar order** in `_parse_args` — `--help` shows args backwards
4. **Quote remaining `${PYARROW_VERSION}` / `${PILLOW_VERSION}`** in `devel_env_setup.sh` for consistency with the `TORCH_TAG` fix

File: jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh (hunk 3 above)
Comment (jiridanek) [reply jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh:53]: cc @coderabbitai

File: jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh (hunk 4 above)
Comment (jiridanek) [reply jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh:69]: cc @coderabbitai

File: runtimes/datascience/ubi9-python-3.12/Dockerfile.cpu (hunk 5 above)
Comment (jiridanek) [reply runtimes/datascience/ubi9-python-3.12/Dockerfile.cpu:68]: cc @coderabbitai

File: tests/test_native_build_versions.py (hunk 6 above)
Comment (jiridanek) [reply tests/test_native_build_versions.py:47]: cc @coderabbitai

File: tests/test_native_build_versions.py (hunk 7 above)
Comment (jiridanek) [reply tests/test_native_build_versions.py]: cc @coderabbitai

File: scripts/pylock_version.py (hunk 8 above)
Comment (jiridanek) [reply scripts/pylock_version.py:169]: cc @coderabbitai

File: jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh (hunk 3 above)
Comment (jiridanek) [reply jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh:53]: Addressed in e9e299c87 — added comments above the torch and accelerate `--platform x86_64` lookups explaining that the pylock pins use x86_64 markers while we build/install the same version on ppc64le.

File: jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh (hunk 4 above)
Comment (jiridanek) [reply jupyter/trustyai/ubi9-python-3.12/devel_env_setup.sh:69]: Addressed in e9e299c87 — quoted `apache-arrow-${PYARROW_VERSION}` and `"${PILLOW_VERSION}"` in the git clone `-b` args.

File: runtimes/datascience/ubi9-python-3.12/Dockerfile.cpu (hunk 5 above)
Comment (jiridanek) [reply runtimes/datascience/ubi9-python-3.12/Dockerfile.cpu:68]: Addressed in e9e299c87 — added `# NB: unquoted heredoc — escape $ for vars that must expand at login` above the `PROFILE_EOF` heredoc in both runtime datascience Dockerfiles.

File: tests/test_native_build_versions.py (hunk 6 above)
Comment (jiridanek) [reply tests/test_native_build_versions.py:47]: Addressed in e9e299c87 — dropped `re.DOTALL` and tightened the pattern to `\bpylock_version\.py\s+(?:\S+\s+)?{package}\b` so it only matches the resolver invocation.

File: tests/test_native_build_versions.py (hunk 7 above)
Comment (jiridanek) [reply tests/test_native_build_versions.py]: Addressed in e9e299c87 — removed the `pyarrow_ppc64le == pyarrow_s390x` equality check. The test now only asserts that each arch resolves a non-empty pin independently.

File: scripts/pylock_version.py (hunk 8 above)
Comment (jiridanek) [reply scripts/pylock_version.py:169]: Addressed in e9e299c87 — switched to `metavar="ARG"` so `--help` no longer shows the positionals backwards. The help text still documents the `PACKAGE` / `PYLOCK PACKAGE` contract.

File: tests/test_native_build_versions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jiridanek) [reply tests/test_native_build_versions.py]: Addressed in e9e299c87 — same fix as the human review: dropped `re.DOTALL` and matched only the `pylock_version.py` invocation line.

File: tests/test_native_build_versions.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+ROOT = Path(__file__).resolve().parents[1]
+
+NATIVE_IMAGE_DIRS = (
+    ROOT / "jupyter/datascience/ubi9-python-3.12",
+    ROOT / "runtimes/datascience/ubi9-python-3.12",
+)
+
+
+@pytest.mark.parametrize("image_dir", NATIVE_IMAGE_DIRS, ids=lambda p: p.relative_to(ROOT).as_posix())
+def test_dockerfiles_read_native_versions_from_pylock(image_dir: Path) -> None:
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply tests/test_native_build_versions.py]: Skipping Allure metadata here — `@allure` is used for container/e2e tests under `tests/containers/`, not for unit tests under `tests/unit/` / top-level script tests. Happy to revisit if we start tagging unit tests with Allure.

**Comment by jiridanek** [comment]
Addressed the open review items in e9e299c87:

1. Comments for intentional `--platform x86_64` lookups (torch/accelerate) inside the ppc64le block
2. Quoted `${PYARROW_VERSION}` / `${PILLOW_VERSION}` git refs
3. NB comment on the expanding `ppc64le.sh` profile heredoc
4. Dropped `re.DOTALL` from Dockerfile pylock lookup regex
5. Removed cross-arch pyarrow equality assertion
6. Fixed CLI metavar (`metavar="ARG"`) so `--help` display matches the contract

Skipped CodeRabbit’s Allure metadata suggestion for these unit/script tests (Allure is used in container e2e tests).

@coderabbitai recheck

**Comment by jiridanek** [comment]
Addressed the open review items on [#2519](https://github.com/red-hat-data-services/notebooks/pull/2519) in `e9e299c87`.

### Fixed
| Feedback | Change |
|----------|--------|
| Comment `--platform x86_64` in ppc64le block | Comments above torch + accelerate lookups |
| Quote `PYARROW_VERSION` / `PILLOW_VERSION` | Quoted git `-b` args |
| Expanding profile heredoc trap | `# NB: unquoted heredoc…` in both runtime Dockerfiles |
| Drop `re.DOTALL` | Tightened Dockerfile lookup regex |
| Cross-arch pyarrow equality | Removed; resolve each arch independently |
| Backwards CLI metavar | `metavar="ARG"` |

### Skipped (with reply)
- **Allure metadata** on unit/script tests — Allure is for `tests/containers/`, not these
- **subtests** nit — low value given short loops + clear assert messages

### Already resolved earlier
- `--format` / shell formatting consistency
- `_ALLOWED_NAMES` aligned with `marker_env` (+ `os_name`)

Replies posted on each thread + a summary comment asking `@coderabbitai recheck`. Tests: **20 passed**.

File: tests/test_native_build_versions.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+
+@pytest.mark.parametrize("image_dir", NATIVE_IMAGE_DIRS, ids=lambda p: p.relative_to(ROOT).as_posix())
+def test_pylock_pins_resolve_for_native_arches(image_dir: Path, pylock_version) -> None:
+    pylock = image_dir / "pylock.toml"
+    locked_version = pylock_version.locked_version
+    # Resolve per-arch independently; versions may diverge across arches in future lockfiles.
+    assert locked_version(pylock, "onnx", platform_machine="ppc64le")
+    assert locked_version(pylock, "pyarrow", platform_machine="ppc64le")
+    assert locked_version(pylock, "pyarrow", platform_machine="s390x")
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply tests/test_native_build_versions.py]: @coderabbitai recheck, fixed

File: semgrep.yaml [hunk 11]
```diff
@@ -1868,7 +1868,9 @@ rules:
       Remediation: Always quote variables in file operations:
         rm "$FILE"         # correct
         rm $FILE           # dangerous
-    pattern-regex: '(rm|cp|mv|eval|chmod|chown|kill|pkill)\s+[^|;]*(?<!["''\\])\$(?:\{[A-Za-z_][A-Za-z0-9_]*(?:[:\-\+\?=][^}]*)?\}|[A-Za-z_][A-Za-z0-9_]*)'
+    # Word boundaries avoid matching substrings (e.g. "rm" in "--platform").
+    # [^\n|;]* keeps the match on one logical command line (no cross-line spans).
+    pattern-regex: '(?<![\w-])(rm|cp|mv|eval|chmod|chown|kill|pkill)(?![\w-])[ \t]+[^\n|;]*(?<!["''\\])\$(?:\{[A-Za-z_][A-Za-z0-9_]*(?:[:\-\+\?=][^}]*)?\}|[A-Za-z_][A-Za-z0-9_]*)'
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply semgrep.yaml]: @coderabbitai recheck, fixed

**Comment by jiridanek** [comment]
## Konflux — [#2519](https://github.com/red-hat-data-services/notebooks/pull/2519) @ `429c811d3`

| Outcome | Pipelines |
|--------|-----------|
| **Pass** | runtime-minimal, runtime-pytorch-cuda, runtime-pytorch-llmcompressor, runtime-tensorflow-cuda, codeserver |
| **Fail (this PR)** | **jupyter-datascience**, **runtime-datascience** |
| **Fail (unrelated)** | jupyter-minimal / pytorch / tensorflow / trustyai (cpu+cuda+rocm) |
| **Timeout 4h** | runtime-pytorch-rocm, runtime-tensorflow-rocm |

**Datascience (real, PR-owned):** ppc64le `onnx-builder` → `pip install -r requirements-min.txt` builds `numpy==1.26.0` from sdist → `No BLAS library detected!` / `openblas found: NO`. Same on jupyter + runtime. That is exactly what [#2534](https://github.com/red-hat-data-services/notebooks/pull/2534) fixes.

**Everything else:** `install_pdf_deps.sh` can’t fetch TexLive from `ftp.math.utah.edu` — curl **28** (timeout) or **60** (SSL issuer). Not caused by the pylock/ONNX changes. ROCm runtimes hit the 4h PipelineRun limit.

---

## Konflux — [#2534](https://github.com/red-hat-data-services/notebooks/pull/2534) @ `94f1b11df`

**No PipelineRuns for the current tip** (KubeArchive + live empty). Last `/build-konflux` was before the restack; tip never got a Konflux rebuild.

GHA subscription builds that did run fail the same TexLive curl path on amd64 (`install_pdf_deps.sh` exit 28); ppc64le/s390x jobs hit ~6h timeouts — not a signal on openblas yet.

---

**Bottom line:** The only stack-owned Konflux break on #2519 is datascience onnx/numpy needing OpenBLAS on ppc64le. Comment `/build-konflux` on #2534 (and optionally re-run datascience-only on #2519 after merge) to confirm the fix; ignore/retrigger the TexLive and ROCm timeout noise.

**Review by ide-developer — APPROVED** [review]

**Comment by jiridanek** [comment]
@coderabbitai review PR comments and whether they have been satisfactorily addressed

File: tests/with_subtests.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+                    f(item)
+            else:
+                # Quiet: only open a subtest context when the body fails, so
+                # successful items do not emit SUBPASSED noise.
+                try:
+                    f(item)
+                except Exception:
+                    with subtests.test(msg=msg, item=item, **fixed_kw):
+                        # Re-raise inside subtest so pytest captures and continues.
+                        raise
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply tests/with_subtests.py:49]: @coderabbitai recheck, fixed

File: tests/unit/test_subtests.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
+"""Coverage for tests.with_subtests.with_subtests."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from tests.with_subtests import with_subtests
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply tests/unit/test_subtests.py:7]: no, this is valid 3.14 exception syntax
